Multiarch enablement on Dockerfile and Dockerfile.build

aleskandro opened the pull request:
- Switching build image to golang:1.17.8
- Changes golangci-lint installation due to golangci/golangci-lint#2374
- Building windows/mac binaries only when architecture is amd64 in fortio Dockerfile
- The docker Debian repo is now added based on the arch

An alternative that allows installing golang-ci-lint as before also on ARM64 is by bumping to golang:1.18. See the issue above.

cc @yselkowitz

Refers ARMOCP-293

aleskandro committed:
Enabling multiarch on Dockerfile and Dockerfile.build

- Switching build image to golang:1.17.8
- Changes golangci-lint installation due to golangci/golangci-lint#2374
- Building windows/mac binaries only when architecture is amd64 in fortio Dockerfile
- Docker repo is now architecture-aware

ldemailly commented:
Thanks for this!

> Changes golangci-lint installation due to https://github.com/golangci/golangci-lint/issues/2374

I scanned the issue and it seem to relate to go 1.18 not 1.17?

Are you sure https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh doesn't work? 
(do you have a sample ci/example of error from your fork?)


In general I prefer ubuntu to debian/alpine/... but I guess it doesn't matter too much if in the end the from scratch images perform and are still small

I don't see this plugged in into the actual https://github.com/fortio/fortio/blob/master/release/release.sh though or 
https://github.com/fortio/fortio/blob/master/.github/workflows/main.yml#L38-L43

to create multiple images or is that somehow automatic and will create multiple images selectable by arch?
(I'm actually not familiar with multi arch docker if you have a faq/doc pointer for me)

aleskandro committed:
CircleCI

aleskandro commented:
Hello @ldemailly, 

> 
> I scanned the issue and it seem to relate to go 1.18 not 1.17?
> 
> Are you sure https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh doesn't work?
> (do you have a sample ci/example of error from your fork?)

building with go 1.17.z and using the previous script fails, at least, on ARM. We could either bump to go1.18 and use the script to install golang-ci-lint, or we can keep on go1.17 but using the `go install` way for that. I tried to be minimal in changes. What do you think? See error below.

> In general I prefer ubuntu to debian/alpine/... but I guess it doesn't matter too much if in the end the from scratch images perform and are still small.

I did it to not delegate here the responsibility to keep [these lines](https://github.com/docker-library/golang/blob/9f431d40c96988dd4302eee93789fa6f071354d2/1.17/bullseye/Dockerfile#L25) in sync for each architecture.


> I don't see this plugged in into the actual https://github.com/fortio/fortio/blob/master/release/release.sh though or
> https://github.com/fortio/fortio/blob/master/.github/workflows/main.yml#L38-L43

> to create multiple images or is that somehow automatic and will create multiple images selectable by arch?
(I'm actually not familiar with multi arch docker if you have a faq/doc pointer for me)


I'm pushing right now another commit that also enables the CI for multi-arch by using Github actions as you are doing already. Note that Github actions only provide x86_64 servers to run the jobs. Hence, the only way I got to make it architecture-aware is by using docker buildx, and cross-building capabilities from the golang compiler and qemu.

Here is the summary of the changes:

    - Updated Dockerfiles for cross-building release artifacts
    - fpm packages by arch
    - fixes Dockerfile.build due to Docker-ce not being available for s390x
    - Makefile actions now use docker buildx
    - Removing pulling of ubuntu:focal in Makefile: docker pull would not cache other architecture's images
    - Windows and MacOS release artifacts are now built in release/Dockerfile.

A successful run of the CI is available at https://github.com/aleskandro/fortio/runs/5818686964?check_suite_focus=true

**I needed to update the .build image manually before. Also, the release took a lot time more to build for each arch.**

### The error from golangci-lint

```
 => [auth] library/golang:pull token for registry-1.docker.io                                                                                        0.0s
 => [linux/arm64  2/11] RUN apt-get -y update &&   apt-get --no-install-recommends -y upgrade &&   DEBIAN_FRONTEND=noninteractive apt-get --no-in  143.0s
 => [linux/arm64  3/11] RUN gem install --no-document fpm                                                                                           20.8s
 => [linux/arm64  4/11] RUN go version # check it's indeed the version we expect                                                                     0.4s
 => [linux/arm64  5/11] RUN curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b /go/bin             22.4s
 => ERROR [linux/arm64  6/11] RUN golangci-lint version                                                                                             56.2s
 => CANCELED [linux/amd64  2/11] RUN apt-get -y update &&   apt-get --no-install-recommends -y upgrade &&   DEBIAN_FRONTEND=noninteractive apt-get   0.0s
------
 > [linux/arm64  6/11] RUN golangci-lint version:
#0 55.68 panic: load embedded ruleguard rules: rules/rules.go:13: can't load fmt
#0 55.68 
#0 55.68 goroutine 1 [running]:
#0 55.68 github.com/go-critic/go-critic/checkers.init.22()
#0 55.68        github.com/go-critic/go-critic@v0.6.2/checkers/embedded_rules.go:46 +0x488
------
Dockerfile.build:17
--------------------
  15 |     # RUN go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.44.2
  16 |     RUN curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $GOPATH/bin
  17 | >>> RUN golangci-lint version
  18 |     
  19 |     # Docker:
--------------------
error: failed to solve: process "/bin/sh -c golangci-lint version" did not complete successfully: exit code: 2

```

aleskandro committed:
Multi-arch integration for release and CI builds by GitHub actions

- Updated Dockerfiles for cross-building release artifacts
- fpm packages by arch
- fixes Dockerfile.build due to Docker-ce not available for s390x
- Makefile actions now uses docker buildx
- Removing pulling of ubuntu:focal in Makefile: docker pull would not cache other architecture's images
- Windows and MacOS release artifacts are now built in release/Dockerfile.in

odidev commented:
I am working on adding arm64 jobs in CircleCI in the consul repo, which uses Fortio images for envoy tests. These changes are working perfectly for building the arm64 image for Fortio and successfully tested it as well.

It will be helpful if Fortio images are released.

ldemailly commented:
I haven't forgotten, it's just been very busy

ldemailly commented on release/Dockerfile.in:
```diff
@@ -1,16 +1,43 @@
+# Linux per-architecture binaries building
+RUN sh -c \
+    'set -ex; for arch in ${archs}; do \
+       make -C fortio official-build-version BUILD_DIR=/build GOARCH=${arch} OFFICIAL_BIN=/tmp/fortio_${arch}; \
```
Assuming it is at all possible, we shouldn't have to rebuild here, the Dockerfile in parent dir should have built everything before and the per architecture from scratch image should have the binary for each arch?
and we should be able to 
COPY --from=release /usr/bin/fortio usr/bin/fortio to get the binary from each of the multi arch?

aleskandro replied:
Hello, I apologize for the late reply. 

It cannot be done in a multiarch scenario this way.

When you build a multiarch image, you build N_ARCH images that will be packaged into a manifest list.

When building this way, copying a file from a previous stage means doing N_ARCH (copy) layers in N_ARCH separate "contexts/containers".

I know we are rebuilding two times the same binaries, but it is done:
1. to build a multiarch manifest list
2. Then, to (cross)-build 4 binaries in a single amd64 image and publish them in the release section of this repository

The alternative could be starting from the second point and building the multi-arch image later. I didn't do to not change more than what I already did till now, your previous workflow. Moreover, it seems a not-so-good workaround to me.

ldemailly replied:
Thanks, I need to read up a bit about multiarch

So it's not possible to address specifically or enumerate all the multi arch images built (to copy from)?

I appreciate greatly the contribution btw! it will be great to have, just trying to wrap my head around the effect on CI 

I probably need to take over your branch/fork to build the new build image and bump the version
( https://github.com/fortio/fortio/tree/master/release#how-to-change-the-build-image )

ldemailly replied:
we'll follow up; if at all possible in #550

ldemailly commented on .github/workflows/main.yml:
```diff
@@ -52,54 +63,9 @@ jobs:
-          asset_path: release/fortio_${{ steps.build.outputs.VERSION }}.orig.tar.gz
-          asset_name: fortio_${{ steps.build.outputs.VERSION }}.orig.tar.gz
-          asset_content_type: application/x-compressed-tar
-# Todo find a way to upload all in 1... it's silly to have to do 1 by 1
```
glad you found how to do that! thx!

ldemailly commented:
I got your branch locally and was trying to make v39 of the build image:
```
$ make -n update-build-image
/Applications/Xcode.app/Contents/Developer/usr/bin/make docker-push-internal IMAGE=.build TAG=v39
./cert-gen
echo "### Now building docker.io/fortio/fortio.build:v39"
docker buildx build --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x -f Dockerfile.build --load -t docker.io/fortio/fortio.build:v39 .
echo "### Now pushing docker.io/fortio/fortio.build:v39"
docker buildx build --push --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x -f Dockerfile.build -t docker.io/fortio/fortio.build:v39 .
```
so I tried
```
$ docker buildx build --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x -f Dockerfile.build --load -t docker.io/fortio/fortio.build:v39 .
[+] Building 0.0s (0/0)                                                                                                                                                                                                    
error: docker exporter does not currently support exporting manifest lists
```

reading around seems --load doesn't quite work in multi arch so I tried
```
$ docker buildx build --platform linux/amd64,linux/arm64,linux/ppc64le,linux/s390x -f Dockerfile.build -t docker.io/fortio/fortio.build:v39 .
[+] Building 0.0s (0/0)                                                                                                                                                                                                    
error: multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")
```

ldemailly committed:
Merge branch 'master' into multiarch

ldemailly committed:
prep for v39, bump golang version, install binary golangci-lint and ref more direct bug

ldemailly commented:
> Hello @ldemailly,
> 
> > I scanned the issue and it seem to relate to go 1.18 not 1.17?
> > Are you sure https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh doesn't work?
> > (do you have a sample ci/example of error from your fork?)
> 
> building with go 1.17.z and using the previous script fails, at least, on ARM. We could either bump to go1.18 and use the script to install golang-ci-lint, or we can keep on go1.17 but using the `go install` way for that. I tried to be minimal in changes. What do you think? See error below.
> 

I reproed indeed the arm failure and found what I think is a more direct issue https://github.com/golangci/golangci-lint/issues/2673 (not related to 1.18) - I updated in https://github.com/fortio/fortio/pull/538/commits/18a6c2c0599fdc46cd1c2b6e2f33bd33e425c330
(can still use the binary install that they recommend)

ldemailly committed:
use v39, multi arch build base. added docker buildx create --use but --load doesn't work for me... had to split build and push and call just push

ldemailly commented on Makefile:
```diff
@@ -96,24 +97,28 @@ FILES_WITH_IMAGE:= .circleci/config.yml Dockerfile Dockerfile.echosrv \
 	@echo "### Now building $(DOCKER_TAG)"
-	docker build -f Dockerfile$(IMAGE) -t $(DOCKER_TAG) .
+	# --load doesn't work on mac m1 with docker desktop even after docker buildx create --use - thoughts?
+	docker buildx build --platform $(BUILDX_PLATFORMS) -f Dockerfile$(IMAGE) --load -t $(DOCKER_TAG) .
```
@aleskandro any idea about why I get 
```
error: docker exporter does not currently support exporting manifest lists
```
If I try to use --load ?
(mac m1 docker desktop)
```
$ docker buildx ls
NAME/NODE             DRIVER/ENDPOINT             STATUS  PLATFORMS
priceless_lovelace *  docker-container                    
  priceless_lovelace0 unix:///var/run/docker.sock running linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
desktop-linux         docker                              
  desktop-linux       desktop-linux               running linux/arm64, linux/amd64, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/arm/v7, linux/arm/v6
default               docker                              
  default             default                     running linux/arm64, linux/amd64, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/arm/v7, linux/arm/v6
```

aleskandro replied:
Hello, yes, --load is not supported when multiple platforms are passed through --platform.

As we need to load the amd64 image in the docker engine registry of the CI, I added a check to verify how many target architectures we are going to build. If only 1, we can use --load, otherwise we can't (See https://github.com/docker/buildx/issues/59). 

A bit worried, instead, about the buildx create command that you added.
Should we consider the builder is already created by a user that wants to build? Beyond the builder, based on the Host OS, you'd have more requirements.

E.g., on Fedora, with docker-ce:
```sh
sudo -i
# Install docker CE
dnf install -y yum-utils
yum-config-manager --add-repo https://download.docker.com/linux/fedora/docker-ce.repo
dnf install -y docker-ce docker-ce-cli containerd.io 
gpasswd -a $SUDO_USER docker
systemctl enable docker && systemctl start docker

# Enable multi-arch support and buildx
dnf install -y qemu-user-static
cat /proc/sys/fs/binfmt_misc/qemu-aarch64 # Smoke check
exit
# Re-login could be needed to get permission for the $SUDO_USER and perform the next commands
docker buildx create --name mybuilder
docker buildx use mybuilder
```

Should we add some info like this in the readme?

aleskandro replied:
edited the comment above

ldemailly replied:
Probably could edit release/README.md with that info

At the moment I'm the one manually pushing the .build using the make targets but yes it's probably not correct to keep making a context (I was wondering if there is a sort of create or update context...)

ldemailly replied:
also it's odd I kept checking and didn't get notifications from github that there were updates on the mr

ldemailly replied:
we'll follow up in #551

ldemailly commented:
looks like it succeeded:

I tagged v1.27.0-pre1 on a copy of your fork in this repo and:

https://github.com/fortio/fortio/runs/6056074667?check_suite_focus=true

->
https://github.com/fortio/fortio/releases/tag/v1.27.0-pre1

->
https://hub.docker.com/layers/203407103/fortio/fortio/1.27.0-pre1/images/sha256-2ec55524db9a0d6cc6ccf50c2dad3bd19a0eccb0224c99027c68fb5cdfeb59b4?context=repo


So beside my question above, if you compare with
https://github.com/fortio/fortio/releases/tag/v1.26.0

it's missing the rpm, the deb and orig.tar.gz

ldemailly committed:
add missing artifacts

ldemailly commented:
> it's missing the rpm, the deb and orig.tar.gz

fixed that part

ldemailly commented:
https://github.com/fortio/fortio/releases/tag/v1.27.0-pre2 has a lot of artifacts now :)

ldemailly committed:
prep for 1.27.0

ldemailly committed:
use latest circleci docker (but should move away from circle anyway)

aleskandro committed:
Adds --load option only when a single platform is set in BUILDX_PLATFORMS

docker/buildx#59
